### PR TITLE
[KSHC] Kyuubi-Spark-Hive-Connector support Apache Spark 3.4

### PR DESCRIPTION
**Pull request by Yikf**
### _Why are the changes needed?_

This pr amins to make KSHC support Apache Spark 3.4. 

- KSHC support Apache Spark 3.4
- Make Apache kyuubi `codecov` module contain the spark-3.4 profile. so that Apache kyubbi CI can cover some modules.

### _How was this patch tested?_
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.readthedocs.io/en/master/contributing/code/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: dev/kyuubi-codecov/pom.xml [hunk 1]
```diff
@@ -209,5 +209,16 @@
                 </dependency>
             </dependencies>
         </profile>
+        <profile>
+            <id>spark-3.4</id>
+            <dependencies>
+                <dependency>
+                    <groupId>org.apache.kyuubi</groupId>
+                    <artifactId>kyuubi-spark-connector-kudu_${scala.binary.version}</artifactId>
```
Comment (pan3793) [inline dev/kyuubi-codecov/pom.xml:217]: actually, I'm going to remove this dummy module ...

File: dev/kyuubi-codecov/pom.xml [hunk 2]
```diff
@@ -209,5 +209,16 @@
...
+                    <version>${project.version}</version>
+                </dependency>
+                <!-- TODO Support Apache Spark 3.4 for KSHC -->
```
Comment (pan3793) [inline dev/kyuubi-codecov/pom.xml:220]: let's fix and add this one instead

**Comment by Yikf** [comment]
Thanks pan @pan3793  for your points, Let me:
- Removes codecov's reference to kudu in profile spark-3.4
- Make KSHC support Spark 3.4 in this patch and add it to the codecov Spark-3.4 profile

By the way, if you don't mind, why did you remove the kudu module?

**Comment by pan3793** [comment]
> By the way, if you don't mind, why did you remove the kudu module?

it's a dummy module, was planned but actually delayed.

**Comment by Yikf** [comment]
Could you please take a look again if you find a time ~ @pan3793

File: extensions/spark/kyuubi-spark-connector-hive/src/main/spark-3.4/org/apache/kyuubi/spark/connector/hive/PartitionedFileUtils.scala [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kyuubi.spark.connector.hive
+
+import org.apache.spark.sql.execution.datasources.PartitionedFile
+
+object PartitionedFileUtils {
```
Comment (pan3793) [inline extensions/spark/kyuubi-spark-connector-hive/src/main/spark-3.4/org/apache/kyuubi/spark/connector/hive/PartitionedFileUtils.scala:22]: we don't need to introduce spark-3.3 and spark-3.4 folders, please use reflection to resolve it.

BTW, it's expected that the connector compiled against spark-3.3 can work on both 3.3 and 3.4

Comment (Yikf) [reply extensions/spark/kyuubi-spark-connector-hive/src/main/spark-3.4/org/apache/kyuubi/spark/connector/hive/PartitionedFileUtils.scala]: Thanks, updated

File: pom.xml [hunk 4]
```diff
@@ -1889,9 +1889,9 @@
                         </java>
                         <scala>
                             <includes>
-                                <include>src/main/scala/**/*.scala</include>
-                                <include>src/test/scala/**/*.scala</include>
-                                <include>src/test/gen/scala/**/*.scala</include>
+                                <include>src/main/**/*.scala</include>
+                                <include>src/test/**/*.scala</include>
+                                <include>src/test/gen/**/*.scala</include>
```
Comment (pan3793) [inline pom.xml:1894]:
```suggestion
                                <include>src/main/scala/**/*.scala</include>
                                <include>src/test/scala/**/*.scala</include>
                                <include>src/test/gen/scala/**/*.scala</include>
```

File: extensions/spark/kyuubi-spark-connector-hive/src/main/scala/org/apache/kyuubi/spark/connector/hive/HiveConnectorUtils.scala [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+import org.apache.spark.sql.types.{ArrayType, MapType, StructField, StructType}
+
+import org.apache.kyuubi.spark.connector.common.SparkUtils
+import org.apache.kyuubi.util.SemanticVersion
+import org.apache.kyuubi.util.reflect.ReflectUtils.invokeAs
+
+object HiveConnectorUtils extends Logging {
+
+  def partitionedFilePath(file: PartitionedFile): String = {
+    if (SparkUtils.isSparkVersionAtLeast("3.4.0")) {
```
Comment (pan3793) [inline extensions/spark/kyuubi-spark-connector-hive/src/main/scala/org/apache/kyuubi/spark/connector/hive/HiveConnectorUtils.scala:38]:
```suggestion
    if (SparkUtils.isSparkVersionAtLeast("3.4")) {
```

Comment (pan3793) [inline extensions/spark/kyuubi-spark-connector-hive/src/main/scala/org/apache/kyuubi/spark/connector/hive/HiveConnectorUtils.scala:38]:
```suggestion
    if (SparkUtils.isSparkVersionAtLeast("3.4")) {
```

File: extensions/spark/kyuubi-spark-connector-hive/src/main/scala/org/apache/kyuubi/spark/connector/hive/HiveConnectorUtils.scala [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+import org.apache.kyuubi.spark.connector.common.SparkUtils
+import org.apache.kyuubi.util.SemanticVersion
+import org.apache.kyuubi.util.reflect.ReflectUtils.invokeAs
+
+object HiveConnectorUtils extends Logging {
+
+  def partitionedFilePath(file: PartitionedFile): String = {
+    if (SparkUtils.isSparkVersionAtLeast("3.4")) {
+      invokeAs[String](file, "urlEncodedPath")
+    } else if (SparkUtils.isSparkVersionAtLeast("3.3.0")) {
```
Comment (pan3793) [inline extensions/spark/kyuubi-spark-connector-hive/src/main/scala/org/apache/kyuubi/spark/connector/hive/HiveConnectorUtils.scala:40]:
```suggestion
    } else if (SparkUtils.isSparkVersionAtLeast("3.3")) {
```

File: extensions/spark/kyuubi-spark-connector-hive/src/main/scala/org/apache/kyuubi/spark/connector/hive/HiveConnectorUtils.scala [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+
+object HiveConnectorUtils extends Logging {
+
+  def partitionedFilePath(file: PartitionedFile): String = {
+    if (SparkUtils.isSparkVersionAtLeast("3.4")) {
+      invokeAs[String](file, "urlEncodedPath")
+    } else if (SparkUtils.isSparkVersionAtLeast("3.3")) {
+      invokeAs[String](file, "filePath")
+    } else {
+      throw KyuubiHiveConnectorException(s"Spark version ${SemanticVersion(SPARK_VERSION)} " +
```
Comment (pan3793) [inline extensions/spark/kyuubi-spark-connector-hive/src/main/scala/org/apache/kyuubi/spark/connector/hive/HiveConnectorUtils.scala:43]:
```suggestion
      throw KyuubiHiveConnectorException(s"Spark version $SPARK_VERSION " +
```

File: extensions/spark/kyuubi-spark-connector-hive/src/main/scala/org/apache/kyuubi/spark/connector/hive/HiveConnectorUtils.scala [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+            field =>
+              Some(StructField(rename.newName, field.dataType, field.nullable, field.metadata)))
+
+        case update: UpdateColumnType =>
+          replace(
+            schema,
+            update.fieldNames,
+            field => {
+              Some(field.copy(dataType = update.newDataType))
+            })
```
Comment (pan3793) [inline extensions/spark/kyuubi-spark-connector-hive/src/main/scala/org/apache/kyuubi/spark/connector/hive/HiveConnectorUtils.scala:117]:
```suggestion
            field => Some(field.copy(dataType = update.newDataType)))
```

File: extensions/spark/kyuubi-spark-connector-hive/src/main/scala/org/apache/kyuubi/spark/connector/hive/HiveConnectorUtils.scala [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+            update.fieldNames,
+            field => Some(field.copy(dataType = update.newDataType)))
+
+        case update: UpdateColumnNullability =>
+          replace(
+            schema,
+            update.fieldNames,
+            field => {
+              Some(field.copy(nullable = update.nullable))
+            })
```
Comment (pan3793) [inline extensions/spark/kyuubi-spark-connector-hive/src/main/scala/org/apache/kyuubi/spark/connector/hive/HiveConnectorUtils.scala:123]:
```suggestion
            field => Some(field.copy(nullable = update.nullable)))
```

File: extensions/spark/kyuubi-spark-connector-hive/src/main/scala/org/apache/kyuubi/spark/connector/hive/HiveConnectorUtils.scala [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+            schema,
+            update.fieldNames,
+            field => Some(field.copy(nullable = update.nullable)))
+
+        case update: UpdateColumnComment =>
+          replace(
+            schema,
+            update.fieldNames,
+            field =>
+              Some(field.withComment(update.newComment)))
```
Comment (pan3793) [inline extensions/spark/kyuubi-spark-connector-hive/src/main/scala/org/apache/kyuubi/spark/connector/hive/HiveConnectorUtils.scala:128]:
```suggestion
            field => Some(field.withComment(update.newComment)))
```

**Review by pan3793 — APPROVED** [review]

File: extensions/spark/kyuubi-spark-connector-hive/src/main/scala/org/apache/kyuubi/spark/connector/hive/HiveConnectorUtils.scala [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+import org.apache.spark.sql.catalyst.catalog.{CatalogTable, CatalogTablePartition}
+import org.apache.spark.sql.connector.catalog.TableChange
+import org.apache.spark.sql.connector.catalog.TableChange.{AddColumn, After, ColumnPosition, DeleteColumn, First, RenameColumn, UpdateColumnComment, UpdateColumnNullability, UpdateColumnPosition, UpdateColumnType}
+import org.apache.spark.sql.execution.command.CommandUtils
+import org.apache.spark.sql.execution.command.CommandUtils.{calculateMultipleLocationSizes, calculateSingleLocationSize}
+import org.apache.spark.sql.execution.datasources.PartitionedFile
+import org.apache.spark.sql.types.{ArrayType, MapType, StructField, StructType}
+
+import org.apache.kyuubi.spark.connector.common.SparkUtils
+import org.apache.kyuubi.util.SemanticVersion
```
Comment (pan3793) [inline extensions/spark/kyuubi-spark-connector-hive/src/main/scala/org/apache/kyuubi/spark/connector/hive/HiveConnectorUtils.scala:32]:
```suggestion
```

**Comment by Yikf** [comment]
Ahh, Thanks @pan3793 for helping me fix these points.

**Review by ulysses-you — APPROVED** [review]

**Comment by pan3793** [comment]
Thanks, merged to master

**Pull request closed without merging**